packages/vite-node/src/client.ts

```diff
@@ -303,6 +303,9 @@ export class ViteNodeRunner {
       env,
       filename: __filename,
       dirname: __dirname,
+      // this requires users to enable "--experimental-import-meta-resolve" even on the latest NodeJS since it uses 2nd argument `parent`.
+      // `import.meta.resolve` becomes `undefined` for vite-node cjs build. So this won't cause syntax error on cjs.
```

Comment:
I don't think that's completely accurate: `'resolve' in import.meta` will resolve different values in Node and Vitest, right?

Reply:
For latest Node (from v20.0.6 https://nodejs.org/docs/latest-v20.x/api/esm.html#importmetaresolvespecifier), `import.meta.resolve` is always defined and `--experimental-import-meta-resolve` only switches 2nd argument `parent` feature. But for older node, `--experimental-import-meta-resolve` is required to have `import.meta.resolve` defined.

So, `'resolve' in import.meta` should be safe for the latest Node (regardless of `--experimental-import-meta-resolve`), but that's not the case for old ones.

I just put next to filename/dirname here, but it's probably better to define this only when `import.meta.resolve` is available.

Reply:
Or maybe you were asking some enumerable property kind of difference?

For now, I updated it to define `import.meta.resolve` by checking it first https://github.com/vitest-dev/vitest/pull/5188/commits/74ca77883f24f23b5435e96880e4533297c9e16a

Reply:
> Or maybe you were asking some enumerable property kind of difference?

Just defining with an `if` is fine for me. I do have another question tho - shouldn't it give an error on the new Node.js version when you don't pass down a flag because it doesn't support `parent`? And all relative URL are resolved relative to the vite-node file.

Reply:
> shouldn't it give an error on the new Node.js version when you don't pass down a flag because it doesn't support parent? And all relative URL are resolved relative to the vite-node file.

Yes, that's very much desired, but I don't know how to do it. I thought about just checking `execArgv`, but the flag might also come from `NODE_OPTIONS`. Do you know some trick for this?

Another way would be to detect feature during runtime by actually testing `import.meta.resolve`, something like:

```ts
// this would give ".../dummy/dummy" if the 2nd argument `parent` is supported 
import.meta.resolve("./dummy", new URL("./dummy", import.meta.url))
```

Reply:
But how is it desired? I would expect this to work on the latest Node.js:

```ts
import.meta.resolve('./dumy')
```

With vite-node resolve it relative to the vite-node file instead of the current file, correct?

Reply:
> With vite-node it is resolved relative to the vite-node file instead of the current file, correct?

When users didn't pass `--experimental-import-meta-resolve`, that's going to happen silently, so that's why I thought we need to detect that first.

If `--experimental-import-meta-resolve` is passed, then the we can override the default parent value, so `import.meta.resolve('./dumy')` should work same both on Node and vite-node.

I updated to check if the 2nd argument feature is available by actually running this inside vite-node

```ts
import.meta.resolve('.', new URL('./__test_parent__/', import.meta.url))
```

Reply:
> Do you only want to support it if there is a flag?

Yes, that's what I was thinking. But, I totally get that exposing this feature only with `--experimental-import-meta-resolve` wouldn't be that good. I'm okay with that putting this PR on hold for now.

Reply:
We already bundle `import-meta-resolve` package, we can just use it here 😄

Reply:
I wrote it in https://github.com/vitejs/vite/discussions/15871#discussioncomment-8436740 but there are some known differences with a polyfill https://github.com/wooorm/import-meta-resolve?tab=readme-ov-file#differences-to-node

It totally depends on the use case, but my feeling is that if users really want to integrate `import.meta.resolve` feature in NodeJs app/library, then they would use it together with custom loader or custom conditions etc...
(actually I had this use case on my own tiny typescript import/export checker https://github.com/hi-ogawa/js-utils/tree/main/packages/icheck-ts)

If that's not the case, then they would've used `createRequire + require.resolve` (for node_modules) or `new URL("./...", ...)` (for relative path) or maybe polyfill by themselves.

I guess this feature request came for the first time, so I think we can wait a little to determine the use cases better. (I didn't even get the response from OP yet...)